### PR TITLE
[WIP] Decouple CUDA

### PR DESCRIPTION
**Pull request by nicolasvasilache**
This is still WIP.

I started going down the rabbit hole of splitting cuda behavior from the rest.
In particular splitting the cuda-specific options into a CudaMappingOptionsProto.
Atm things do not run because of improper behavior with 
```MappingOptions::MappingOptionsProto proto``` and 
```CudaMappingOptions::CudaMappingOptionsProto proto``` that collide.
I'll pick it up again tomorrow, finish and cleanup (in particular, bogus fixCPU compilation commit), but if someone wants to take a look in the meantime and make some WIP progress, be my guest.

It is unclear yet how to split the large commit.

### CODE REVIEW

File: src/core/polyhedral/mapped_scop.cc [hunk 1]
```diff
@@ -572,6 +573,9 @@ std::tuple<std::string, tc::Grid, tc::Block> MappedScop::codegen(
   code << "extern \"C\" {" << std::endl
        << emitCudaKernel(specializedName, *mappedScopForCodegen) << "}"
        << std::endl;
+#else
+  code << "TC was compiled with -DWITH_CUDA=0 which disabled cuda compilation";
+#endif
```
Comment (skimo-openhub) [inline src/core/polyhedral/mapped_scop.cc]: The next commit _reverts_ these changes.
If they are not needed, then don't apply them first.
Btw, I seem to be able to compile with `-DWITH_CUDA=0` without these changes.

File: src/core/CMakeLists.txt [hunk 2]
```diff
@@ -27,8 +26,12 @@ add_library(
   polyhedral/schedule_print.cc
   polyhedral/scop.cc
   polyhedral/separation.cc
-  polyhedral/tighten_launch_bounds.cc
   polyhedral/unroll.cc
+
+  # Builds and runs without the CUDA SDK and used in cross-compilation tests
+  polyhedral/codegen_cuda.cc
+  polyhedral/cuda/cuda_mapping_types.cc
+  polyhedral/cuda/cuda_tighten_launch_bounds.cc
```
Comment (skimo-openhub) [inline src/core/CMakeLists.txt]: This looks like an independent change.

**Comment by skimo-openhub** [comment]
Can you add some motivation to your commit messages?
It's difficult to review the commits without knowing their purpose.

File: include/tc/core/mapping_options-inl.h [hunk 3]
```diff
@@ -418,4 +226,65 @@ MappingOptions& MappingOptions::intraTileSchedulePositiveOrthant(bool b) {
   return *this;
 }
 
+//
+// Predefined stratgies
+//
+template <typename ConcreteMappingOptionsType>
+ConcreteMappingOptionsType MappingOptions::makeUnmappedMappingOptions() {
```
Comment (skimo-openhub) [inline include/tc/core/mapping_options-inl.h]: Why did you have to move this strategies?
Also, typo in "stratgies".

File: include/tc/core/mapping_options.h [hunk 4]
```diff
@@ -333,9 +230,14 @@ class SchedulerOptions : public SchedulerOptionsView {
 /// block sizes).  Provides static constructors for common operator options.
 /// Provides fluent (chainable) API for progressively modifying the options.
 class MappingOptions {
- private:
+ public:
+  virtual void concretize() = 0; // Make this class abstract
```
Comment (skimo-openhub) [inline include/tc/core/mapping_options.h]: Hmm... couldn't you just make the destructor abstact?
Actually, `toProtobufSerializedString` is already abstact.

File: include/tc/core/mapping_options.h [hunk 5]
```diff
@@ -367,29 +269,8 @@ class MappingOptions {
   template <typename... Args>
   MappingOptions& tile(Args...);
 
-  inline MappingOptions& mapToThreads(std::initializer_list<uint64_t> threads);
-  inline MappingOptions& mapToThreads(
-      uint64_t x,
-      uint64_t y = CudaDimView::defaultDim,
-      uint64_t z = CudaDimView::defaultDim);
-  inline MappingOptions& mapToThreads(const std::vector<uint64_t>& threads);
-  MappingOptions& mapToThreads(const std::string& commaSeparatedSizes);
```
Comment (skimo-openhub) [inline include/tc/core/mapping_options.h]: Update the MappingOptions documentation to the removal of these properties.

File: include/tc/core/mapping_options_cpp_printer.h [hunk 6]
```diff
@@ -45,7 +45,7 @@ class MappingOptionsCppPrinter {
       MappingOptionsCppPrinter& prn,
       const MappingOptions& options);
 
- private:
+ public:
```
Comment (skimo-openhub) [inline include/tc/core/mapping_options_cpp_printer.h]: Why not `protected`?

File: src/proto/mapping_options.proto [hunk 7]
```diff
@@ -43,24 +43,42 @@ message MappingOptionsProto {
   // Tile sizes.  If empty or not provided, do not tile.
   optional TilingProto tiling = 4;
   // Block dimensions (min 1, max 3 values).
-  required CudaDimProto block = 5;
+  // required CudaDimProto block = 5;
   // Grid dimensions (min 1, max 3 values).
-  required CudaDimProto grid = 6;
+  // required CudaDimProto grid = 6;
```
Comment (skimo-openhub) [inline src/proto/mapping_options.proto]: Remove commented out "code".
Use `reserved` instead (if allowed by the version of protobuf we are using).

File: test/test_execution_engine_cache.cc [hunk 8]
```diff
@@ -45,8 +45,10 @@ TEST(ATenCompilationCacheTest, Matmul) {
   at::Tensor b = at::CUDA(at::kFloat).rand({4, 5});
   std::vector<at::Tensor> inputs = {a, b};
   std::vector<at::Tensor> outputs;
-  auto mappingOptions = tc::MappingOptions::makeMlpMappingOptions();
-  auto handle = atCompl.compile("matmul", inputs, mappingOptions);
+  auto mappingOptions = tc::CudaMappingOptions::makeMlpCudaMappingOptions();
+  tc::CudaMappingOptions mappingOptions2(
+      mappingOptions.toProtobufSerializedString());
```
Comment (skimo-openhub) [inline test/test_execution_engine_cache.cc]: Why do we need `mappingOptions2`?

**Comment by nicolasvasilache** [comment]
Not ready yet will reopen when ready

**Pull request closed without merging**